README.md
## rosflight (melodic) - 1.3.1-1

The packages in the `rosflight` repository were released into the `melodic` distro by running `/usr/bin/bloom-release rosflight --rosdistro melodic` on `Fri, 27 Mar 2020 19:23:56 -0000`

These packages were released:
- `rosflight`
- `rosflight_firmware`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_sim`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: https://github.com/rosflight/rosflight-release.git
- rosdistro version: `1.3.0-1`
- old version: `1.3.0-1`
- new version: `1.3.1-1`

Versions of tools used:

- bloom version: `0.9.3`
- catkin_pkg version: `0.4.16`
- rosdep version: `0.18.0`
- rosdistro version: `0.8.0`
- vcstools version: `0.1.42`


## rosflight (kinetic) - 1.3.1-1

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release rosflight --rosdistro kinetic` on `Fri, 27 Mar 2020 19:15:29 -0000`

These packages were released:
- `rosflight`
- `rosflight_firmware`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_sim`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: https://github.com/rosflight/rosflight-release.git
- rosdistro version: `1.3.0-1`
- old version: `1.3.0-1`
- new version: `1.3.1-1`

Versions of tools used:

- bloom version: `0.9.3`
- catkin_pkg version: `0.4.16`
- rosdep version: `0.18.0`
- rosdistro version: `0.8.0`
- vcstools version: `0.1.42`


## rosflight (melodic) - 1.3.0-1

The packages in the `rosflight` repository were released into the `melodic` distro by running `/usr/bin/bloom-release rosflight --track melodic --rosdistro melodic --new-track` on `Thu, 19 Mar 2020 06:27:28 -0000`

These packages were released:
- `rosflight`
- `rosflight_firmware`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_sim`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.3.0-1`

Versions of tools used:

- bloom version: `0.9.3`
- catkin_pkg version: `0.4.16`
- rosdep version: `0.18.0`
- rosdistro version: `0.8.0`
- vcstools version: `0.1.42`


## rosflight (kinetic) - 1.3.0-1

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release rosflight --rosdistro kinetic` on `Thu, 19 Mar 2020 06:06:15 -0000`

These packages were released:
- `rosflight`
- `rosflight_firmware`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_sim`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: https://github.com/rosflight/rosflight-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-1`
- new version: `1.3.0-1`

Versions of tools used:

- bloom version: `0.9.3`
- catkin_pkg version: `0.4.16`
- rosdep version: `0.18.0`
- rosdistro version: `0.8.0`
- vcstools version: `0.1.42`


## rosflight (lunar) - 1.0.0-0

The packages in the `rosflight` repository were released into the `lunar` distro by running `/usr/bin/bloom-release rosflight --track lunar --rosdistro lunar --new-track` on `Thu, 15 Mar 2018 21:18:25 -0000`

These packages were released:
- `rosflight`
- `rosflight_firmware`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_sim`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.0-0`

Versions of tools used:

- bloom version: `0.6.2`
- catkin_pkg version: `0.4.1`
- rosdep version: `0.12.1`
- rosdistro version: `0.6.8`
- vcstools version: `0.1.40`


## rosflight (kinetic) - 1.0.0-1

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release rosflight --track kinetic --rosdistro kinetic` on `Wed, 14 Mar 2018 22:07:41 -0000`

These packages were released:
- `rosflight`
- `rosflight_firmware`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_sim`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: https://github.com/rosflight/rosflight-release.git
- rosdistro version: `1.0.0-0`
- old version: `1.0.0-0`
- new version: `1.0.0-1`

Versions of tools used:

- bloom version: `0.6.2`
- catkin_pkg version: `0.4.1`
- rosdep version: `0.12.1`
- rosdistro version: `0.6.8`
- vcstools version: `0.1.40`


## rosflight (kinetic) - 1.0.0-0

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release rosflight --track kinetic --rosdistro kinetic` on `Tue, 13 Mar 2018 16:26:16 -0000`

These packages were released:
- `rosflight`
- `rosflight_firmware`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_sim`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: https://github.com/rosflight/rosflight-release.git
- rosdistro version: `0.1.3-0`
- old version: `0.1.3-0`
- new version: `1.0.0-0`

Versions of tools used:

- bloom version: `0.6.2`
- catkin_pkg version: `0.4.1`
- rosdep version: `0.12.1`
- rosdistro version: `0.6.6`
- vcstools version: `0.1.40`


## rosflight (kinetic) - 0.1.3-0

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release rosflight --track kinetic --rosdistro kinetic` on `Sat, 03 Jun 2017 05:51:33 -0000`

These packages were released:
- `rosflight`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: https://github.com/rosflight/rosflight-release.git
- rosdistro version: `0.1.2-0`
- old version: `0.1.2-0`
- new version: `0.1.3-0`

Versions of tools used:

- bloom version: `0.5.26`
- catkin_pkg version: `0.3.1`
- rosdep version: `0.11.5`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## rosflight (kinetic) - 0.1.2-0

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release rosflight --track kinetic --rosdistro kinetic` on `Wed, 24 May 2017 18:24:56 -0000`

These packages were released:
- `rosflight`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: https://github.com/rosflight/rosflight-release.git
- rosdistro version: `0.1.1-0`
- old version: `0.1.1-0`
- new version: `0.1.2-0`

Versions of tools used:

- bloom version: `0.5.26`
- catkin_pkg version: `0.3.1`
- rosdep version: `0.11.5`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## rosflight (kinetic) - 0.1.1-0

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release rosflight --track kinetic --rosdistro kinetic` on `Wed, 24 May 2017 15:56:01 -0000`

These packages were released:
- `rosflight`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: https://github.com/rosflight/rosflight-release.git
- rosdistro version: `0.1.0-1`
- old version: `0.1.0-1`
- new version: `0.1.1-0`

Versions of tools used:

- bloom version: `0.5.26`
- catkin_pkg version: `0.3.1`
- rosdep version: `0.11.5`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## rosflight (kinetic) - 0.1.0-1

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release --rosdistro kinetic --track kinetic rosflight --edit` on `Wed, 24 May 2017 05:32:29 -0000`

These packages were released:
- `rosflight`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: unknown
- rosdistro version: `null`
- old version: `0.1.0-0`
- new version: `0.1.0-1`

Versions of tools used:

- bloom version: `0.5.26`
- catkin_pkg version: `0.3.1`
- rosdep version: `0.11.5`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


## rosflight (kinetic) - 0.1.0-0

The packages in the `rosflight` repository were released into the `kinetic` distro by running `/usr/bin/bloom-release --rosdistro kinetic --track kinetic rosflight --edit` on `Wed, 24 May 2017 05:17:43 -0000`

These packages were released:
- `rosflight`
- `rosflight_msgs`
- `rosflight_pkgs`
- `rosflight_utils`

Version of package(s) in repository `rosflight`:

- upstream repository: https://github.com/rosflight/rosflight.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `0.1.0-0`

Versions of tools used:

- bloom version: `0.5.26`
- catkin_pkg version: `0.3.1`
- rosdep version: `0.11.5`
- rosdistro version: `0.6.2`
- vcstools version: `0.1.39`


# rosflight-release
Release repository for the ROSflight ROS stack


debian/changelog
ros-melodic-rosflight-pkgs (1.3.1-1stretch) stretch; urgency=high



 -- Daniel Koch <daniel.p.koch@gmail.com>  Fri, 27 Mar 2020 06:00:00 -0000

ros-melodic-rosflight-pkgs (1.3.0-1stretch) stretch; urgency=high

  * Contributors: Cameron McQuinn, Jacob Willis

 -- Daniel Koch <daniel.p.koch@gmail.com>  Thu, 19 Mar 2020 06:00:00 -0000

ros-melodic-rosflight-pkgs (1.0.0-1stretch) stretch; urgency=high

  * Small updates to package.xml's and readme
  * Contributors: Daniel Koch, James Jackson

 -- Daniel Koch <daniel.p.koch@gmail.com>  Tue, 13 Mar 2018 06:00:00 -0000

ros-melodic-rosflight-pkgs (0.1.3-1stretch) stretch; urgency=high

  * Updated package.xml files
  * Contributors: Daniel Koch

 -- Daniel Koch <daniel.p.koch@gmail.com>  Fri, 02 Jun 2017 06:00:00 -0000

ros-melodic-rosflight-pkgs (0.1.2-1stretch) stretch; urgency=high



 -- Daniel Koch <daniel.p.koch@gmail.com>  Wed, 24 May 2017 06:00:00 -0000

ros-melodic-rosflight-pkgs (0.1.1-1stretch) stretch; urgency=high



 -- Daniel Koch <daniel.p.koch@gmail.com>  Wed, 24 May 2017 06:00:00 -0000

ros-melodic-rosflight-pkgs (0.1.0-1stretch) stretch; urgency=high

  * Added rosflight_pkgs metapackage
  * Contributors: Daniel Koch

 -- Daniel Koch <daniel.p.koch@gmail.com>  Mon, 22 May 2017 06:00:00 -0000




rosflight_firmware/firmware/boards/naze/lib/breezystm32/main.c
/*
   main.c : entry routine for for STM32F103CB

   Adapted from https://github.com/multiwii/baseflight/blob/master/src/main.c

   This file is part of BreezySTM32.

   BreezySTM32 is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   BreezySTM32 is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with BreezySTM32.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "breezystm32.h"

serialPort_t * Serial1;

extern void SetSysClock(bool overclock);

static void _putc(void *p, char c)
{
    (void)p; // avoid compiler warning about unused variable
    serialWrite(Serial1, c);

    while (!isSerialTransmitBufferEmpty(Serial1));
}

int main(void)
{
    // Configure clock, this figures out HSE for hardware autodetect
    SetSysClock(0);

    systemInit();

    Serial1 = uartOpen(USART1, NULL, 115200, MODE_RXTX);

    setup();

    init_printf( NULL, _putc);
   
    while (1)
        loop();
}


rosflight_firmware/firmware/boards/naze/lib/breezystm32/drv_i2c.h
/*
   drv_i2c.h :  I^2C support for STM32F103CB

   Adapted from https://github.com/multiwii/baseflight/blob/master/src/drv_i2c.h

   This file is part of BreezySTM32.

   BreezySTM32 is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   BreezySTM32 is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with BreezySTM32.  If not, see <http://www.gnu.org/licenses/>.
 */

#pragma once

typedef enum I2CDevice {
    I2CDEV_1,
    I2CDEV_2,
    I2CDEV_MAX = I2CDEV_2
} I2CDevice;

typedef enum {
    READ,
    WRITE
} i2cJobType_t;

enum {
    I2C_JOB_DEFAULT,
    I2C_JOB_QUEUED,
    I2C_JOB_BUSY,
    I2C_JOB_COMPLETE,
    I2C_JOB_ERROR
};

typedef struct i2cJob{
    i2cJobType_t type;
    uint8_t addr;
    uint8_t reg;
    uint8_t* data;
    uint8_t length;
    struct i2cJob* next_job;
    volatile uint8_t* status;
    void (*CB)(void);
} i2cJob_t;

#define I2C_BUFFER_SIZE 16

void i2cInit(I2CDevice index);
uint16_t i2cGetErrorCounter(void);

// Blocking I2C functions (returns value success or failure)
bool i2cWriteBuffer(uint8_t addr_, uint8_t reg_, uint8_t len_, uint8_t *data);
bool i2cWrite(uint8_t addr_, uint8_t reg, uint8_t data);
bool i2cRead(uint8_t addr_, uint8_t reg, uint8_t len, uint8_t *buf);

// ===================================================================
// Asynchronous I2C handler
// To use this, queue up a job, and create a callback you want called when the job is finished
// You can track progress of the job using the status pointer.  Otherwise, functions the same
// as the blocking versions.
//
// This uses a circular buffer to stage jobs for the I2C peripheral.  The buffer is, by default, 16 jobs
// long (I2C_BUFFER_SIZE), with a maximum size of 256. I hope you never queue up that many jobs, because
// that will take a long time to process However, if you were to reach the limit, it would then start
// ignoring new jobs until there was space on the buffer.
//
// For an example of how to use, check out mpu6050_request_read_temp - the non-blocking way to read
// the accelerometer
void i2c_queue_job(i2cJobType_t type, uint8_t addr_, uint8_t reg_, uint8_t *data, uint8_t length, volatile uint8_t *status_, void (*CB)(void));


rosflight_firmware/firmware/boards/naze/lib/breezystm32/drv_ms5611.h
/*
   drv_ms5611.h : driver for Measurement Specialties MS5611 barometer

   Adapted from https://github.com/multiwii/baseflight/blob/master/src/drv_ms5611.h

   This file is part of BreezySTM32.

   BreezySTM32 is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   BreezySTM32 is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with BreezySTM32.  If not, see <http://www.gnu.org/licenses/>.
 */

#pragma once

bool ms5611_init(void);

// blocking I2C update function
void ms5611_update(void);
void ms5611_read(float* pressure, float* temperature);

// Asynchronous ms5611 functions
bool ms5611_present(void);
void ms5611_async_update(void);
void ms5611_async_read(float* pressure, float* temperature);


rosflight_firmware/firmware/boards/naze/lib/breezystm32/drv_hmc5883l.h
/*
   drv_hmc5883l.h :  Support for HMC5883L Magnetometer

   Adapted from https://github.com/multiwii/baseflight/blob/master/src/drv_hmc5883l.h

   This file is part of BreezySTM32.

   BreezySTM32 is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   BreezySTM32 is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with BreezySTM32.  If not, see <http://www.gnu.org/licenses/>.
 */


#pragma once

bool hmc5883lInit(int boardVersion);

// Blocking I2C Read Method
void hmc5883l_update();
void hmc5883l_read(int16_t *magData);

// Asynchronous I2C method
void hmc5883l_request_async_update();
void hmc5883l_async_read(int16_t *magData);

bool hmc5883l_present();


rosflight_firmware/firmware/boards/naze/lib/breezystm32/drv_ms5611.c
/*
   drv_ms5611.c : driver for Measurement Specialties MS5611 barometer

   Adapted from https://github.com/multiwii/baseflight/blob/master/src/drv_ms5611.c

   This file is part of BreezySTM32.

   BreezySTM32 is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   BreezySTM32 is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with BreezySTM32.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <breezystm32.h>
#include <limits.h>

// MS5611, Standard address 0x77
#define MS5611_ADDR             0x77

#define CMD_RESET               0x1E // ADC reset command
#define CMD_ADC_READ            0x00 // ADC read command
#define CMD_ADC_CONV            0x40 // ADC conversion command
#define CMD_ADC_D1              0x00 // ADC D1 conversion
#define CMD_ADC_D2              0x10 // ADC D2 conversion
#define CMD_ADC_256             0x00 // ADC OSR=256
#define CMD_ADC_512             0x02 // ADC OSR=512
#define CMD_ADC_1024            0x04 // ADC OSR=1024
#define CMD_ADC_2048            0x06 // ADC OSR=2048
#define CMD_ADC_4096            0x08 // ADC OSR=4096
#define CMD_PROM_RD             0xA0 // Prom read command
#define PROM_NB                 8

static uint32_t ms5611_ut;  // static result of temperature measurement
static uint32_t ms5611_up;  // static result of pressure measurement
static uint16_t ms5611_c[PROM_NB];  // on-chip ROM
static uint8_t  ms5611_osr = CMD_ADC_4096;
float pressure = 0.0;
float temperature = 0.0;
static uint8_t init_state = 0;

static void ms5611_reset(void)
{
  i2cWrite(MS5611_ADDR, CMD_RESET, 1);
  delayMicroseconds(2800);
}

static uint16_t ms5611_prom(int8_t coef_num)
{
  uint8_t rxbuf[2] = { 0, 0 };
  i2cRead(MS5611_ADDR, CMD_PROM_RD + coef_num * 2, 2, rxbuf); // send PROM READ command
  return rxbuf[0] << 8 | rxbuf[1];
}

int8_t ms5611_crc(uint16_t *prom)
{
  int32_t i, j;
  uint32_t res = 0;
  uint8_t crc = prom[7] & 0xF;
  prom[7] &= 0xFF00;

  bool blankEeprom = true;

  for (i = 0; i < 16; i++) {
    if (prom[i >> 1]) {
      blankEeprom = false;
    }
    if (i & 1)
      res ^= ((prom[i >> 1]) & 0x00FF);
    else
      res ^= (prom[i >> 1] >> 8);
    for (j = 8; j > 0; j--) {
      if (res & 0x8000)
        res ^= 0x1800;
      res <<= 1;
    }
  }
  prom[7] |= crc;
  if (!blankEeprom && crc == ((res >> 12) & 0xF))
    return 0;

  return -1;
}

static uint32_t ms5611_read_adc(void)
{
  uint8_t rxbuf[3];
  i2cRead(MS5611_ADDR, CMD_ADC_READ, 3, rxbuf); // read ADC
  return (rxbuf[0] << 16) | (rxbuf[1] << 8) | rxbuf[2];
}

static void ms5611_start_ut(void)
{
  i2cWrite(MS5611_ADDR, CMD_ADC_CONV + CMD_ADC_D2 + ms5611_osr, 1); // D2 (temperature) conversion start!
}

static void ms5611_get_ut(void)
{
  ms5611_ut = ms5611_read_adc();
}

static void ms5611_start_up(void)
{
  i2cWrite(MS5611_ADDR, CMD_ADC_CONV + CMD_ADC_D1 + ms5611_osr, 1); // D1 (pressure) conversion start!
}

static void ms5611_get_up(void)
{
  ms5611_up = ms5611_read_adc();
}

static void ms5611_calculate()
{
  int32_t press = 0;
  int64_t temp = 0;
  int64_t delt = 0;
  if(ms5611_up > 9085466 * 2 / 3 && ms5611_ut > 0)
  {
    int64_t dT = (int64_t)ms5611_ut - ((int64_t)ms5611_c[5] << 8);
    int64_t off = ((int64_t)ms5611_c[2] << 16) + (((int64_t)ms5611_c[4] * dT) >> 7);
    int64_t sens = ((int64_t)ms5611_c[1] << 15) + (((int64_t)ms5611_c[3] * dT) >> 8);
    temp = 2000 + ((dT * (int64_t)ms5611_c[6]) >> 23);

    if (temp < 2000) { // temperature lower than 20degC
      delt = temp - 2000;
      delt = 5 * delt * delt;
      off -= delt >> 1;
      sens -= delt >> 2;
      if (temp < -1500) { // temperature lower than -15degC
        delt = temp + 1500;
        delt = delt * delt;
        off -= 7 * delt;
        sens -= (11 * delt) >> 1;
      }
      temp -= ((dT * dT) >> 31);
    }
    press = ((((uint64_t)ms5611_up * sens) >> 21) - off) >> 15;

    pressure = (float)press; // Pa
    temperature = (float)temp/ 100.0 + 273.0; // K

//    float pre_adjust_altitude = fast_alt(pressure);

//    altitude = fast_alt(pressure) - offset;
  }
}


// =======================================================================================

bool ms5611_init(void)
{
  bool ack = false;
  uint8_t sig;
  int i;

  while(millis() < 10); // No idea how long the chip takes to power-up, but let's make it 10ms

  ack = i2cRead(MS5611_ADDR, CMD_PROM_RD, 1, &sig);
  if (!ack)
    return false;
  else
    init_state = 1;

  ms5611_reset();

  init_state = 2;

  // read all coefficients
  for (i = 0; i < PROM_NB; i++)
    ms5611_c[i] = ms5611_prom(i);
  // check crc, bail out if wrong
  if (ms5611_crc(ms5611_c) != 0)
    return false;

  return true;
}

void ms5611_update(void)
{
  static uint32_t next_time_ms = 0;
  static int state = 0;

  if(millis() > next_time_ms)
  {
    if (state)
    {
      ms5611_get_up();
      ms5611_start_ut();
      next_time_ms += 10;
      state = 0;
    }
    else
    {
      ms5611_get_ut();
      ms5611_start_up();
      state = 1;
      next_time_ms += 10;
      ms5611_calculate();
    }
  }
}

void ms5611_read(float* press, float* temp)
{
  (*press) = pressure;
  (*temp) = temperature;
}


//===================================================================
// ASYNC FUNCTIONS
static uint8_t pressure_buffer[3];
static uint8_t temp_buffer[3];

static uint8_t temp_command = 1;
static uint8_t pressure_command = 1;
static volatile uint8_t temp_start_status = 0;
static volatile uint8_t temp_read_status = 0;
static volatile uint8_t pressure_read_status = 0;
static volatile uint8_t pressure_start_status = 0;
static volatile uint8_t baro_state = 0;
static volatile uint32_t next_update_ms = 0;

static volatile uint8_t init_status;
static uint8_t init_command;

void ms5611_init_CB(void)
{
  // we successfully found the sensor
  if (init_status == I2C_JOB_COMPLETE)
  {
    init_state ++;
    next_update_ms = millis() + 30;
  }
}

void temp_request_CB(void)
{
  next_update_ms = millis() + 15;
}

void pressure_request_CB(void)
{
  next_update_ms = millis() + 15;
}

void pressure_read_CB(void)
{
  ms5611_up = (pressure_buffer[0] << 16) | (pressure_buffer[1] << 8) | pressure_buffer[2];
}

static void temp_read_CB(void)
{
  ms5611_ut = (temp_buffer[0] << 16) | (temp_buffer[1] << 8) | temp_buffer[2];
}

void ms5611_async_update(void)
{
  uint32_t now_ms = millis();
  // if it's not time to do anything, just return
  if (now_ms < next_update_ms)
  {
    return;
  }
  else
  {
    // this may be reduced by an i2c callback, but at the very least, monitor at 10 Hz
    next_update_ms += 100;
  }

  // Try to initialize the sensor if we haven't already
  if (init_state  < 2)
  {
    if (init_state == 0)
    {
      init_command = 1;
      i2c_queue_job(READ, MS5611_ADDR, CMD_PROM_RD, &init_command, 1, &init_status, ms5611_init_CB);
      return;
    }
    else if (init_state == 1)
    {
      init_command = 1;
      i2c_queue_job(READ, MS5611_ADDR, CMD_RESET, &init_command, 1, &init_status, ms5611_init_CB);
      return;
    }
  }

  switch (baro_state)
  {
  case 0:
    // Read the pressure
    i2c_queue_job(READ,
                  MS5611_ADDR,
                  CMD_ADC_READ,
                  pressure_buffer,
                  3,
                  &pressure_read_status,
                  &pressure_read_CB);
//    baro_state++;
//    break;
//  case 1:
    // start a temp conversion
    i2c_queue_job(WRITE,
                  MS5611_ADDR,
                  CMD_ADC_CONV + CMD_ADC_D2 + ms5611_osr,
                  &temp_command,
                  1,
                  &temp_start_status,
                  &temp_request_CB);
    baro_state = 1;
    break;
  case 1:
    // Read the temperature
    i2c_queue_job(READ,
                  MS5611_ADDR,
                  CMD_ADC_READ,
                  temp_buffer,
                  3,
                  &temp_read_status,
                  &temp_read_CB);
//    baro_state = 0;
//    break;
//  case 3:
    // start a pressure conversion
    i2c_queue_job(WRITE,
                  MS5611_ADDR,
                  CMD_ADC_CONV + CMD_ADC_D1 + ms5611_osr,
                  &pressure_command,
                  1,
                  &pressure_start_status,
                  &pressure_request_CB);
    baro_state = 0;
    break;
  default:
    baro_state = 0;
    break;
  }

  ms5611_calculate();
}

bool ms5611_present()
{
  return init_state > 0;
}

void ms5611_async_read(float* press, float* temp)
{
  (*press) = pressure;
  (*temp) = temperature;
}
